src/shims/env.rs

```diff
@@ -91,19 +176,59 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
                     .deallocate(var, None, MiriMemoryKind::Machine.into())?;
             }
             this.update_environ()?;
-            Ok(0)
+            Ok(0) // return zero on success
         } else {
             Ok(-1)
         }
     }
 
+    fn setenvironmentvariablew(
+        &mut self,
+        name_op: OpTy<'tcx, Tag>, // LPCWSTR lpName,
+        value_op: OpTy<'tcx, Tag>, // LPCWSTR lpValue,
+    ) -> InterpResult<'tcx, i32> {
+        let mut this = self.eval_context_mut();
+        this.assert_target_os("windows", "SetEnvironmentVariableW");
+
+        let name_ptr = this.read_scalar(name_op)?.not_undef()?;
+        let value_ptr = this.read_scalar(value_op)?.not_undef()?;
+
+        let mut new = None;
+        if !this.is_null(name_ptr)? {
+            let name = this.read_os_str_from_target_str(name_ptr)?;
+            if this.is_null(value_ptr)? {
+                // Delete environment variable `{name}`
+                if let Some(var) = this.machine.env_vars.map.remove(&name) {
+                    this.memory.deallocate(var, None, MiriMemoryKind::Machine.into())?;
+                    this.update_environ()?;
+                }
+                return Ok(1);  // return non-zero on success
+            }
+            if !name.is_empty() && !name.to_string_lossy().contains('=') {
+                let value = this.read_os_str_from_target_str(value_ptr)?;
+                new = Some((name.to_owned(), value.to_owned()));
+            }
+        }
+        if let Some((name, value)) = new {
+            let var_ptr = alloc_env_var_as_target_str(&name, &value, &mut this)?;
+            if let Some(var) = this.machine.env_vars.map.insert(name, var_ptr) {
+                this.memory
+                    .deallocate(var, None, MiriMemoryKind::Machine.into())?;
+            }
+            this.update_environ()?;
+            Ok(1) // return non-zero on success
+        } else {
+            Ok(0)
+        }
+    }
+
     fn unsetenv(&mut self, name_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
         let this = self.eval_context_mut();
 
         let name_ptr = this.read_scalar(name_op)?.not_undef()?;
         let mut success = None;
         if !this.is_null(name_ptr)? {
-            let name = this.read_os_str_from_c_str(name_ptr)?.to_owned();
+            let name = this.read_os_str_from_target_str(name_ptr)?.to_owned();
```

Comment:
Where is this called on Windows? Shouldn't this rather get the same target_os `assert!` as `getenv` and `setenv`?

Reply:
`unsetenv` is not called in Windows. For deleting an environment variable in Windows, it calls `SetEnvironmentVariableW` with a null pointer to unset an environment variable.

Reply:
Okay then there is no reason to use `target_str` here, and we should have an `assert!`.